source/Detector/Scaler/AMDSScalerDetector.cpp

```diff
@@ -0,0 +1,152 @@
+#include "AMDSScalerDetector.h"
+
+#include <QDebug>
+
+#include "beamline/AMPVControl.h"
+#include "beamline/AMControlSet.h"
+
+#include "DataHolder/AMDSScalarDataHolder.h"
+#include "DataElement/AMDSDataTypeDefinitions.h"
+#include "DataElement/AMDSFlatArray.h"
+
+#include "Detector/Scaler/AMDSScalerConfigurationMap.h"
+
+#include "util/AMErrorMonitor.h"
+
+AMDSScalerDetector::AMDSScalerDetector(AMDSScalerConfigurationMap *scalerConfiguration, QObject *parent)
+    : QObject(parent)
+{
+	scalerConfiguration_ = scalerConfiguration;
+
+	connected_ = false;
+	defaultDwellTime_ = 1; // default 1ms
+	dwellTime_ = 1;
+	defaultScansInABuffer_ = 1000;
+	scansInABuffer_ = 1000;
+	defaultTotalNumberOfScans_ = 0;
+	totalNumberOfScans_ = 0;
+
+	initializePVControls();
+}
+
+AMDSScalerDetector::~AMDSScalerDetector()
+{
+	channelStatusControlSet_->clear();
+	pvControlSet_->clear();
+}
+
+void AMDSScalerDetector::onFetchScanBuffer()
+{
+	if (!connected_)
+		return;
+
+	AMDSDataHolderList scanBufferDataHolderList;
+
+	int channelCount = enabledChannelCount();
+	QVector<int> countBuffer(channelCount * scansInABuffer_ + 1); // need one extra buffer for the # of channels
+	scanControl_->values(channelCount, countBuffer.data());
+
+	// put the counts array to the FlatArrayList
+	for (int scanIndex=0; scanIndex < scansInABuffer_; scanIndex++) {
+		AMDSFlatArray channelCountArray(scalerConfiguration_->dataType(), channelCount);
+
+		int channelScanStartIndex = scanIndex * channelCount + 1;
+		channelCountArray.setValues(countBuffer.mid(channelScanStartIndex, channelCount));
+
+		AMDSLightWeightScalarDataHolder *scalerDataHolder = new AMDSLightWeightScalarDataHolder(scalerConfiguration_->dataType());
+		scalerDataHolder->setData(&channelCountArray);
+		scanBufferDataHolderList.append(scalerDataHolder);
+	}
+
+	if (scanBufferDataHolderList.count() > 0) {
+		emit newScalerScanDataReceived(scanBufferDataHolderList);
+	}
+
+}
+
+void AMDSScalerDetector::onAllControlsConnected(bool connected)
+{
+	if(connected_ != connected)
+		connected_ = connected;
+
+	if (!connected_) {
+		AMErrorMon::alert(this, 0, "PV control is NOT connected.");
+		return;
+	}
+
+	dwellTimeControl_->move(defaultDwellTime_);
+	scansInABufferControl_->move(defaultScansInABuffer_);
+	totalNumberOfScansControl_->move(defaultTotalNumberOfScans_);
+
+	continuousScanControl_->move(AMDSScalerDetector::Continuous); // switch to continuous mode
+}
+
+void AMDSScalerDetector::onAllControlsTimedOut()
+{
+	connected_ = false;
+	AMErrorMon::alert(this, 0, "PV control is NOT connected.");
+}
+
+void AMDSScalerDetector::onDwellTimeControlValueChanged(double newValue)
+{
+	if (dwellTime_ != newValue) {
+		AMErrorMon::alert(this, 0, QString("Scaler %1 dwell time changed from %2ms to %3ms.").arg(scalerName()).arg(dwellTime_).arg(newValue));
+
+		dwellTime_ = (int)newValue;
```

Comment:
This should be casted before the comparison.  That will ensure better non-floating point equivalent comparisons.